app/src/main/java/com/github/polybooks/database/FBInterestDatabase.kt

```diff
@@ -0,0 +1,213 @@
+     */
+    override fun addCourse(course: Course) : CompletableFuture<Course> {
+        val future = CompletableFuture<Course>()
+
+        FirebaseProvider.getFirestore()
+            .collection(courseCollection)
+            .document(course.name)
+            .set(course, SetOptions.merge())
+            .addOnSuccessListener { future.complete(course) }
+            .addOnFailureListener { future.completeExceptionally(DatabaseException("Failed to insert ${course.name} into Database because of: $it")) }
+
+        return future
+    }
+
+    /**
+     * Remove a field
+     * Warning! If it has sub-collections they are currently not deleted
+     * Deleting sub-collections is not implemented yet because we are not using sub-collections here.
+     * A valid argument could also be that interests are very rarely removed
+     * So it could be fine to removed them from the console (automatically deleting all the sub-collections)
+     * Instead of using a function for it.
+     */
+    fun removeField(field: Field) : CompletableFuture<Boolean>  {
+        val future = CompletableFuture<Boolean>()
+
+        FirebaseProvider.getFirestore()
+            .collection(fieldCollection)
+            .document(field.name)
+            .delete()
+            .addOnSuccessListener { future.complete(true) }
+            .addOnFailureListener { future.completeExceptionally(DatabaseException("Could not delete ${field.name} because of: $it")) }
+
+        return future
+    }
+/*
+    /**
+     * Remove a semester and all its potential sub-collections (users and books)
+     */
+    fun removeSemester(semester: Semester) : CompletableFuture<Boolean>  {
+
+    }
+
+    /**
+     * Remove a course and all its potential sub-collections (users and books)
+     */
+    fun removeCourse(course: Course) : CompletableFuture<Boolean>  {
+
+    }
+
+ */
+
+    // TODO maybe add listAllBooksOfCourse(course), etc. and listAllUsersInterestedIn(interest) if relevant
+    // TODO Also addUserInterest() and removeUserInterest() instead of setUserInterest()
+
+    /**
+     * List all the Fields in the database.
+     * */
+    override fun listAllFields(): CompletableFuture<List<Field>> {
+        val future = CompletableFuture<List<Field>>()
+
+        FirebaseProvider.getFirestore()
+            .collection(fieldCollection)
+            .get()
+            .addOnSuccessListener { documentSnapshots ->
+                val fieldsList = documentSnapshots.map { snapshot ->
+                    Field(name = snapshot.get("name") as String)
+                }
+                future.complete(fieldsList)
+            }
+            .addOnFailureListener { future.completeExceptionally(DatabaseException("Could not retrieve the list of all fields because of: $it")) }
+
+        return future
+    }
+
+
+    /**
+     * List all the Semesters in the database.
+     * */
+    override fun listAllSemesters(): CompletableFuture<List<Semester>> {
```

Comment:
Il y a du code duppliqué qui pourrait probablement être factorisé entre listAllSemesters et listAllCourses et listAllFields

Reply:
I thought about that, but this duplicated code is interleaved with parameters relating to which of the 3 types of interest is listed, so if the factorisation needs to include `when()` on every second line, I'm not convinced it helps

Reply:
Ok, I had something similar in the SaleDatabase so I understand.